tests/ttnn/unit_tests/operations/test_moreh_norm.py

```diff
@@ -288,26 +464,97 @@ def test_moreh_norm_backward(input_shape, p, dim_rtol_atol, keepdim, device):
+        p (float): The order of the norm.
+        dim_rtol_atol (list): List containing the dimension(s), relative tolerance, and absolute tolerance for the comparison.
+        compute_kernel_options: Configuration options for the compute kernel.
+        device: The device to run the computation on.
+
+    Raises:
+        AssertionError: If the computed gradients from Tenstorrent's implementation and PyTorch are not close.
+    """
     torch.manual_seed(2024)
     dim, rtol, atol = dim_rtol_atol
     run_moreh_norm_backward(input_shape, p, dim, rtol, atol, device, compute_kernel_options=compute_kernel_options)
 
 
-@pytest.mark.parametrize("p", [1.5], ids=["p=2.0"])
+@pytest.mark.parametrize("p", [2.0])
 @pytest.mark.parametrize(
     "dim_rtol_atol",
     [
         [[], 0.2, 0.2],
+        [None, 0.2, 0.2],
+        [0, 0.1, 0.1],
+        [1, 0.1, 0.1],
+        [2, 0.1, 0.1],
+        [3, 0.1, 0.1],
+        [[0, 1], 0.1, 0.1],
+        [[0, 1, 2], 0.15, 0.15],
+        [[0, 1, 2, 3], 0.2, 0.2],
+        [[0, 1, 3], 0.15, 0.15],
+        [[0, 2, 3], 0.15, 0.15],
+        [[1, 2], 0.1, 0.1],
+        [[1, 2, 3], 0.15, 0.15],
+        [[1, 3], 0.1, 0.1],
+        [[2, 3], 0.1, 0.1],
+    ],
+    ids=[
+        "global_norm(dim=[])",
+        "global_norm(dim=None)",
+        "N",
+        "C",
+        "H",
+        "W",
+        "NC",
+        "NCH",
+        "NCHW",
+        "NCW",
+        "NHW",
+        "CH",
+        "CHW",
+        "CW",
+        "HW",
     ],
 )
 @pytest.mark.parametrize(
     "input_shape",
     [
-        [4, 4],
+        [TILE_HEIGHT, TILE_WIDTH],
+        [2, 2, 2 * TILE_HEIGHT + 13, 2 * TILE_WIDTH + 13],
     ],
 )
 def test_moreh_norm_backward_callback(input_shape, p, dim_rtol_atol, device, use_program_cache):
+    """
+    Test the norm backward operation in Tenstorrent's implementation with and without the program cache.
+    Verifies that the number of program cache entries remains consistent when running backward norm operations.
+
+    Args:
+        input_shape (list of int): Shape of the input tensor.
+        p (float): The order of the norm.
+        dim_rtol_atol (list): List containing the dimension(s), relative tolerance, and absolute tolerance for the comparison.
+        device: The device to run the computation on.
+        use_program_cache: Use the program cache.
+
+    Raises:
+        AssertionError: If the number of program cache entries differs between runs with the same settings.
+    """
     torch.manual_seed(2024)
     dim, rtol, atol = dim_rtol_atol
-    for _ in range(2):
-        run_moreh_norm_backward(input_shape, p, dim, rtol, atol, device)
+    num_program_cache_entries_list = []
+    for i in range(4):
+        if i < 2:
+            run_moreh_norm_backward(input_shape, p, dim, rtol, atol, device, keepdim=False)
+        else:
+            run_moreh_norm_backward(input_shape, p, dim, rtol, atol, device, keepdim=True)
```

Comment:
Could `keepdim` just be a parameter of `test_moreh_norm_backward_callback`? Is there any particular reason it was implemented this way instead?

Reply:
We need to verify that the number of program cache entries remains consistent when running backward norm operations.

The purpose of the test:
- Running the operation the first and second time with the same settings should result in the same number of cache entries.
- On the third run, new cache entries will be created, but the third and fourth runs should still have the same number of cache entries.

If we parameterize the keepdim, changing the setting will reset the number of cache entries to zero.

Reply:
```Python
def test_moreh_norm_backward_callback(..., keepdim):
  for i in range(2):
      run_moreh_norm_backward(keepdim=keepdim)
```
Using the code snippet like above allow you to maintain the same keepdim value across both runs.
Is there any issue with this approach?

Reply:
The existing approach, which your specified above is true, but it's not test the behavior when the cache entries increases for same operation.

Reply:
In fact, I'm skeptical about checking the count. I understand that `cache_entries` refers to the number of caches created, but this value can vary depending on the implementation of the norm, making it not particularly meaningful.

For example, in the case of `moreh_norm`, separate caches are created depending on `keepdim`. However, if `compute_program_hash` is implemented correctly, the number of cache_entries should not increase.

Looking at the implementation of `moreh_norm`, the same kernel is used regardless of `keepdim`.

Reply:
It is indeed a change. Since keepdim is an attribute of operation_attributes_t, it will likely result in a cache miss in theory.

The experiment confirms this as well:
```bash
 tests.ttnn.unit_tests.operations.test_moreh_norm:test_moreh_norm_callback:383 - num_program_cache_entries_list=[1, 1, 2, 2]
 tests.ttnn.unit_tests.operations.test_moreh_norm:test_moreh_norm_callback:383 - num_program_cache_entries_list=[1, 1, 2, 2]
 tests.ttnn.unit_tests.operations.test_moreh_norm:test_moreh_norm_callback:383 - num_program_cache_entries_list=[2, 2, 4, 4]
 tests.ttnn.unit_tests.operations.test_moreh_norm:test_moreh_norm_backward_callback:559 - num_program_cache_entries_list=[1, 1, 2, 2]
 tests.ttnn.unit_tests.operations.test_moreh_norm:test_moreh_norm_backward_callback:559 - num_program_cache_entries_list=[1, 1, 2, 2]
 tests.ttnn.unit_tests.operations.test_moreh_norm:test_moreh_norm_backward_callback:559 - num_program_cache_entries_list=[1, 1, 2, 2]
```

Reply:
The issue is not whether `cache_entries` currently changes or not. 
The current change is likely due to the lack of implementation of `compute_program_hash`, and `cache_entries` may not change depending on how it's implemented.

Reply:
@hschoi4448 We do not care if `cache_entries` changes, we are only checking whether or not re-running the same operation twice would cause a new cache item to generate or not (somewhat testing `compute_program_hash` and indirectly, `override_runtime_arguments`). We expect that under a re-run with a new buffer, cached programs are used and overriden correctly and no new cache items are generated.

Why we are storing and comparing the number of caches like this is literally just to assert that no new cache items are created unless we change a operation attribute (in this case, `keepdim`).

Also fyi `compute_program_hash` already has a default implementation to hash everything:
https://github.com/tenstorrent/tt-metal/blob/8d64e353ea2bee32d4e2ddf6eebb27aa1cb48da0/ttnn/cpp/ttnn/device_operation.hpp#L103-L115

Reply:
In the callback test, the two things that need to be tested are:
 1. Whether a cache hit occurs.
 2. Whether the values are correct.
However, TT does not have a feature to check whether a cache hit has occurred.

So it seems that some tests are substituting this with the number of cache entries, but it's not a meaningful test. 
1. The first reason is that just because cache entries were created, it does not guarantee that a hit occurred
2.  The second reason is that the correct number of cache entries when a cache hit occurs depends on the implementation.

In the case of `moreh_norm`, if the implementation uses the same kernel for both `keepdim` true and false, then the number of cache entries does not need to increase. However, if implemented with separate kernels, the `num_program_cache_entries` would increase.

In the case of complex ops like `nll_loss`, if the op internally calls a `sum` op that is not implemented by it, the `num_program_cache_entries` may return a number that is completely unrelated to its own implementation.

Reply:
>Why we are storing and comparing the number of caches like this is literally just to assert that no new cache items are created unless we change a operation attribute (in this case, keepdim).

When the keepdim changes, whether a new cache is created or not can also depend on the implementation. 

The reason a new cache is currently being generated is that the default `compute_program_hash` is being used, which creates a new cache whenever any values differ, regardless of the kernel implementation. 

This leads to low cache utilization, and it should be addressed in the future by directly implementing `compute_program_hash` to improve efficiency.

Reply:
> In the callback test, the two things that need to be tested are:
> 
>     1. Whether a cache hit occurs.
If a cache hit occured, wouldn't you expect that the `num_program_cache_entries` call to **NOT** increment (keep in mind that call accumulates and isn't reset unless `disable_and_clear_program_cache` is called), so if a hit occurred, it should stay at the same count? (Check https://github.com/tenstorrent/tt-metal/blob/main/tt_metal/impl/device/program_cache.hpp for the implementation)
I'd say if you find a case where re-running the exact same operations with the exact same arguments that generate the exact same program hash, but **does not** result in a cache hit and the counter increase, that might be an issue with the framework itself, don't you think?
 
>     2. Whether the values are correct.
>        However, TT does not have a feature to check whether a cache hit has occurred.
Again, what I said above. Also, if you really wants to, just hook the file I mentioned above and inspect the program cache yourself. It's basically an unordered map, so you can just log whenever a cache hit occured

Reply:
Whatever, I made the changes as Hyungsuk requested. Please check.

Reply:
Also, I updated the test cases table in the description.